csharp/ql/src/semmle/code/csharp/Type.qll

```diff
@@ -797,6 +797,31 @@ class DelegateType extends RefType, Parameterizable, @delegate_type {
   override string getAPrimaryQlClass() { result = "DelegateType" }
 }
 
+/**
+ * A function pointer type, for example
+ *
+ * ```csharp
+ * delegate*<int, void>
+ * ```
+ */
+class FunctionPointerType extends Type, Parameterizable, @function_pointer_type {
+  /** Gets the return type of this function pointer. */
+  Type getReturnType() { function_pointer_return_type(this, getTypeRef(result)) }
+
+  /** Gets the calling convention. */
+  int getCallingConvention() { function_pointer_calling_conventions(this, result) }
```

Comment:
We should wrap the `int` in a class layer that covers the various calling conventions (unless there are just too many?)

Reply:
I think there are 6 options. What do you mean by wrapping the `int`? Do you mean having 6 dedicated predicates, like in `TypeParameterConstraints` for `general_type_parameter_constraints()`?

Reply:
No, what I meant was something like
```ql
private newtype TCallingConvention = MkCallingConvention(int i) { function_pointer_calling_conventions(_, i) }

class CallingConvention extends TCallingConvention { }

// Add sub classes

CallingConvention getCallingConvention() { exists(int i | function_pointer_calling_conventions(this, i) and result = MkCallingConvention(i) }
```